spatialos-sdk/src/worker/alpha_locator.rs

```diff
@@ -0,0 +1,494 @@
+use std::ffi::CString;
+
+use futures::{Async, Future};
+
+use spatialos_sdk_sys::worker::*;
+
+use crate::worker::{internal::utils::cstr_to_string, parameters::ProtocolLoggingParameters};
+
+pub struct AlphaLocator {
+    pub(crate) internal: *mut Worker_Alpha_Locator,
+}
+
+impl AlphaLocator {
+    pub fn create_development_player_identity_token(
+        hostname: &str,
+        port: u16,
+        request: &mut PlayerIdentityTokenRequest,
+    ) -> PlayerIdentityTokenFuture {
+        unsafe {
+            let cstr = CString::new(hostname).unwrap();
+            let (mut params, _data) = request.to_worker_sdk();
+            PlayerIdentityTokenFuture::new(Worker_Alpha_CreateDevelopmentPlayerIdentityTokenAsync(
+                cstr.as_ptr(),
+                port,
+                &mut params as *mut _,
+            ))
+        }
+    }
+
+    pub fn create_development_login_tokens(
+        hostname: &str,
+        port: u16,
+        request: &mut LoginTokensRequest,
+    ) -> LoginTokensFuture {
+        unsafe {
+            let cstr = CString::new(hostname).unwrap();
+            let (mut params, _data) = request.to_worker_sdk();
+            LoginTokensFuture::new(Worker_Alpha_CreateDevelopmentLoginTokensAsync(
+                cstr.as_ptr(),
+                port,
+                &mut params as *mut _,
+            ))
+        }
+    }
+
+    pub fn new(hostname: &str, port: u16, params: &AlphaLocatorParameters) -> Self {
+        let hostname = CString::new(hostname).unwrap();
+        let (cparams, _data) = params.to_worker_sdk();
+
+        unsafe {
+            let ptr = Worker_Alpha_Locator_Create(hostname.as_ptr(), port, &cparams);
+
+            AlphaLocator { internal: ptr }
+        }
+    }
+}
+
+pub struct AlphaLocatorParameters {
+    player_identity: PlayerIdentityCredentials,
+    use_insecure_connection: bool,
+    logging: Option<ProtocolLoggingParameters>,
+}
+
+impl AlphaLocatorParameters {
+    pub fn new(credentials: PlayerIdentityCredentials) -> Self {
+        AlphaLocatorParameters {
+            player_identity: credentials,
+            use_insecure_connection: false,
+            logging: None,
+        }
+    }
+
+    pub fn with_insecure_connection(mut self) -> Self {
+        self.use_insecure_connection = true;
+        self
+    }
+
+    pub fn with_logging(self) -> Self {
+        self.with_logging_params(ProtocolLoggingParameters::default())
+    }
+
+    pub fn with_logging_params(mut self, params: ProtocolLoggingParameters) -> Self {
+        self.logging = Some(params);
+        self
+    }
+
+    fn to_worker_sdk(&self) -> (Worker_Alpha_LocatorParameters, Vec<CString>) {
```

Comment:
Yup - this is another pattern I've used extensively in the FFI. 

I think I can replace these with a wrapper struct that's generic over the data type and enforces a lifetime linked with the corresponding Rust struct. 

Watch this space 😉

Reply:
So I've replaced these with: 

```
pub struct WrappedNativeData<T, U> {
    pub native_data: T,
    pub underlying_data: U,
}
```

to encapsulate the data a bit more. 

Taking the example of: `AlphaLocatorParameters`: for `AlphaLocatorParameters` to own the CString would mean that I would need need to tie the lifetime of `Worker_Alpha_LocatorParameter` to that of the `AlphaLocatorParameter` - but this isn't really possible: 

```
pub struct WrappedStruct<'a, T> {
	pub data: T,
	_marker: PhantomData<&'a ()>
}
```

You can still move `T` out of the struct. I'd need to return a reference to the FFI struct that you construct in the `to_worker_sdk(&self)` method which afaik is not possible. I.e. - 

```
pub struct WrappedNativeData<'a, T, U> {
    pub native_data: &'a T,
    pub underlying_data: U,
}
```
---

One idea I've floated is to `Box` and `std::mem::forget` the CString and manually `impl Drop` for each of the `Worker_*` types which reconstructs the `CString` from the ptr and drops it there.

This feels (and probably is) the _most_ dangerous, but actually correctly ties the lifetime of the `CString` to the FFI struct together.

Reply:
I don't think it's necessary to fully tie the lifetime of `Worker_Alpha_LocatorParameters` to the `AlphaLocatorParameters`, at least not in the Rust sense. This is unsafe code and `Worker_Alpha_LocatorParameters` is a C struct that uses pointers instead of Rust references, so it's not really possible to tie the lifetimes together in the correct Rust-y way.

I think `WrappedNativeData` isn't really necessary here. It's sufficient for `AlphaLocatorParameters` to have a method like `as_worker_sdk()` that returns a `Worker_Alpha_LocatorParameters` setup to point to the data owned by the `AlphaLocatorParameters`. You'd want to have a comment noting that the `Worker_Alpha_LocatorParameters` is borrowing data from `AlphaLocatorParameters` and that the `AlphaLocatorParameters` should therefore be kept alive for however long the C API will still be using that data, but that's about all that we need to do in my opinion. Ultimately, it's not important that `Worker_Alpha_LocatorParameters`'s lifetime is tied to the owned data, since we're never actually using that data from the Rust side (it only exists to be consumed by the C SDK). All that matters is that the data lives as long as the C SDK needs.

As such, I think the simplest thing to do is to have `AlphaLocatorParameters` own the `CString` data directly (i.e. use `CString` internally instead of `String`). That avoids the need of having the intermediate `WrappedNativeData` struct entirely.

---

On a (somewhat) related note, it would be good if there were more documentation in the C API around the ownership and lifetime of any data being passed around as a pointer. For example, how long do the strings in `Worker_Alpha_LocatorParameters` need to live? Does the SDK copy them immediately? Or do they need to stay alive for longer than that? The C API uses pointers for pretty much everything, and while I can usually make an educated guess as to the expected ownership of the data, it would be super helpful to have more concrete information.

Reply:
> have `AlphaLocatorParameters` own the `CString` data directly (i.e. use `CString` internally instead of `String`)

My concern with this is that you leak implementation details to a user. 

This can be offset by hiding the members of the struct and ask users to use a getter/setter style interface.. will need to take a look at existing usage for the feasibility of that.

Reply:
It seems like the members in `AlphaLocatorParameters` are already private, though? Even if that wasn't intended, I think it's reasonable to keep it that way. We still need to enforce the invariant that any user-provided strings don't have a null byte in the middle, which would make more sense to do at the point where the user passes in the string, rather than at the point where we conver the `AlphaLocatorParameters` into C types.